### PR TITLE
chore(medusa): Add transaction on mutation actions

### PR DESCRIPTION
**Pull request by adrien2p**
**What**
Wrap all actions that require the usage of an atomic phase into a transaction from the handler

This pr will stop to the orders (down) to avoid to big pr, i will create a new one for the rest of the handlers https://github.com/medusajs/medusa/pull/1855

### CODE REVIEW

File: packages/medusa/src/api/routes/admin/orders/request-return.ts [hunk 1]
```diff
@@ -89,176 +90,177 @@ export default async (req, res) => {
...
+  await manager.transaction(async (transactionManager) => {
+    let idempotencyKey
+    try {
+      idempotencyKey = await idempotencyKeyService
+        .withTransaction(transactionManager)
+        .initializeRequest(headerKey, req.method, req.params, req.path)
+    } catch (error) {
+      res.status(409).send("Failed to create idempotency key")
+      return
+    }
```
[the remark this answers is not in the document]
Comment (srindom) [reply packages/medusa/src/api/routes/admin/orders/request-return.ts:103]: The `idempotencyKeyService.workStage` creates a transaction so when doing it this way you are creating nested transactions which we don't want.

**Review by srindom — CHANGES_REQUESTED** [review]

File: packages/medusa/src/api/routes/admin/orders/request-return.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (adrien2p) [reply packages/medusa/src/api/routes/admin/orders/request-return.ts:103]: no, the workStage, from what I see, reuse the given transaction and pass it the `func`. Since the manager from the atonicPhase is the one given through the `withTransction`

[the remark this answers is not in the document]
Comment (adrien2p) [reply packages/medusa/src/api/routes/admin/orders/request-return.ts:103]: Why do we want it to happen in a separate transaction than the one currently running?

**Review by pKorsholm — COMMENTED** [review]
Very nice! I have a couple of comments that I have put here more generally:

I think you have missed a `withTransaction` in `draft-orders/register-payment.ts` on `createTaxLines`

question: Do we add transactions to the `delete-metadata` endpoints, because it's not added to `orders/delete-metadata`

We could clean up the transaction in `orders/fulfill-swap` since it retrieves and sets response status and body inside the transaction (same for `orders/process-swap-payments`)

**Comment by adrien2p** [comment]
> Very nice! I have a couple of comments that I have put here more generally:
> 
> I think you have missed a `withTransaction` in `draft-orders/register-payment.ts` on `createTaxLines`
> 
> question: Do we add transactions to the `delete-metadata` endpoints, because it's not added to `orders/delete-metadata`
> 
> We could clean up the transaction in `orders/fulfill-swap` since it retrieves and sets response status and body inside the transaction (same for `orders/process-swap-payments`)

I ve done the changes, thanks for having spotted that 🙏 
For the delete metadata, it is part of the end points that are removed in different PR's

**Review by srindom — APPROVED** [review]
LGTM!
---
_The following is a non-blocking thought that I think we should plan some work around in the future:_

Separately, I am also beginning to realize that some of our work here is not necessarily related to transactions. Namely, what we are trying to avoid is the following:

- Request A calls `SomeService.doSomething` - the `doSomething` method creates a transaction through `atomicPhase` which in turn sets the `SomeService` instance's `manager = transactionManager`.
- When Request B comes in and calls `SomeService.[anyMethod]` we risk using the manager set by Request A instead of a fresh manager. 
- If Request A commits while Request B is still doing work we are in trouble. 

The above problem is removed by wrapping everything in transactions, because `withTransaction` clones the `SomeService` instance and this is the important part. That is, the cloning work (not the transaction work) ensures that no two requests use the same `SomeService` instance.

We would benefit from removing any notion of "state" from all basic services so that they become pure instead. Will note this down for later discussion.

**Comment by adrien2p** [comment]
> LGTM! 🎉
> 
> _The following is a non-blocking thought that I think we should plan some work around in the future:_
> 
> Separately, I am also beginning to realize that some of our work here is not necessarily related to transactions. Namely, what we are trying to avoid is the following:
> 
> * Request A calls `SomeService.doSomething` - the `doSomething` method creates a transaction through `atomicPhase` which in turn sets the `SomeService` instance's `manager = transactionManager`.
> * When Request B comes in and calls `SomeService.[anyMethod]` we risk using the manager set by Request A instead of a fresh manager.
> * If Request A commits while Request B is still doing work we are in trouble.
> 
> The above problem is removed by wrapping everything in transactions, because `withTransaction` clones the `SomeService` instance and this is the important part. That is, the cloning work (not the transaction work) ensures that no two requests use the same `SomeService` instance.
> 
> We would benefit from removing any notion of "state" from all basic services so that they become pure instead. Will note this down for later discussion.

Yes indeed, most of the problems was that the requests was using the same service instance. But we only wraps the necessary bit in a transaction and everything that does not need a transaction is done out of it. What we wanted to avoid, was to set everything to be scoped and work with new instances whenever we need it instead of all the time. 

But we also had a transaction issue since some of the work needed to be done in a transaction but that the request was using an existing transaction instead of a new one.

**Comment by srindom** [comment]
Agree, the main goal should be to make the services pure; let's discuss a few options, but like this suggestion:

> another way would be that all methods that need a transaction, accept a context object that can contains the transactionManager and that can be pass along

File: packages/medusa/src/api/routes/store/carts/create-cart.ts [hunk 2]
```diff
@@ -81,26 +81,24 @@ export default async (req, res) => {
...
+    if (!regions?.length) {
+      throw new MedusaError(
+        MedusaError.Types.INVALID_DATA,
+        `A region is required to create a cart`
+      )
     }
 
-    let cart = await cartService.withTransaction(manager).create({
+    regionId = regions[0].id
+  }
```
Comment (srindom) [inline packages/medusa/src/api/routes/store/carts/create-cart.ts:98]: thought(non-blocking): this is a good candidate for a validator middleware; could be nice with a separate PR for this.

**Review by srindom — APPROVED** [review]
LGTM

File: packages/medusa/src/api/routes/store/carts/create-cart.ts (hunk 2 above)
Comment (adrien2p) [reply packages/medusa/src/api/routes/store/carts/create-cart.ts:98]: yes indeed, but as you proposed, a separate pr would be a better fit I think. Should we create an issue for that?
